tool/lib/license_utils.dart

```diff
@@ -0,0 +1,255 @@
+  }
+
+  /// Returns a copy of the given [file] with the [existingHeader] replaced by
+  /// the [replacementHeader]. Reads and writes the entire file contents all
+  /// at once, so performance may degrade for large files.
+  Future<File> rewriteLicenseHeader(
+    File file,
+    String existingHeader,
+    String replacementHeader,
+  ) async {
+    final File rewrittenFile = File('${file.path}.tmp');
+    // TODO: [mossmana] Convert to using a stream
+    await file.readAsString().then((String contents) async {
+      final String replacementContents = contents.replaceFirst(
+        existingHeader,
+        replacementHeader,
+      );
+      await rewrittenFile.writeAsString(replacementContents);
+    });
+    return rewrittenFile;
+  }
+
+  /// Bulk update license headers for files in the [directory] as configured
+  /// in the [config]. Returns a list of file paths that were updated.
+  Future<Map<String, List<String>>> bulkUpdate(
+      Directory directory, LicenseConfig config,) async {
+    final List<String> includedPaths = [];
+    final List<String> updatedPaths = [];
+    final List<File> files =
+        directory.listSync(recursive: true).whereType<File>().toList();
+    for (final file in files) {
+      if (!config.shouldExclude(file)) {
+        includedPaths.add(file.path);
+        final String extension = p.extension(file.path).replaceAll('.', '');
+        final YamlList removeIndices =
+            config.getRemoveIndicesForExtension(extension);
+        for (final removeIndex in removeIndices) {
+          final String existingLicenseText = config.removeLicenses[removeIndex];
+          final int addIndex = config.getAddIndexForExtension(extension);
+          final String replacementLicenseText = config.addLicenses[addIndex];
+          final int fileLength = await file.length();
+          final int existingLicenseTextLength = existingLicenseText.length;
+          const int buffer = 20;
+          // Assume that the license text will be near the start of the file,
+          // but add in some buffer.
+          int byteCount = existingLicenseTextLength + buffer;
+          byteCount = fileLength > byteCount ? byteCount : fileLength;
+          final Map<String, String> replacementInfo = await getReplacementInfo(
+            file,
+            existingLicenseText,
+            replacementLicenseText,
+            existingLicenseText.length,
+          );
+          final String? existingHeader =
+              replacementInfo[LicenseHeader.existingHeaderKey];
+          final String? replacementHeader =
+              replacementInfo[LicenseHeader.replacementHeaderKey];
+          if (existingHeader != null && replacementHeader != null) {
+            final File rewrittenFile = await rewriteLicenseHeader(
+              file,
+              existingHeader,
+              replacementHeader,
+            );
+            final File backupFile = file.copySync('${file.path}.bak');
```

Comment:
what's the backupFile for? it doesn't look like we do anything with it.

Reply:
This is in the event something bad happens and the changes need to be rolled back. (This utility may be run by someone who hasn't committed code changes yet, for example.)

Reply:
We delete `backupFile` below though, don't we?

Reply:
If there is an exception, the program aborts before getting to the deletion.

Reply:
Maybe over architecting here, but wanted to have some fallback. Is there a better way to handle this case?

Reply:
Do we expect to encounter exceptions? Do you have any examples of cases where we could hit them?

We'd probably be better off including a `--dry-run` to tell users which files are going to be modified without actually modifying them. If you want to be extra thorough, you could check to see if the tree is clean or not and ask for confirmation before running the tool.

Reply:
I like the dry-run idea. I'll go with that and just leave it up to the end user to decide how to proceed. Thanks!